pupil_src/shared_modules/annotations.py

```diff
@@ -7,239 +7,329 @@
-        return csv_keys + tuple(sorted(user_keys))
+    def __init__(self, g_pool, *args, **kwargs):
+        super().__init__(g_pool, *args, **kwargs)
+        self.annotations = self.load_annotations("annotation_player")
+        no_or_empty_player_data_file = len(self.annotations) == 0
+        if no_or_empty_player_data_file:
+            self.annotations = self.load_annotations("annotation")
+        self.last_frame_ts = None
+        self.last_frame_index = -1
+
+    def load_annotations(self, file_name):
+        annotation_pldata = fm.load_pldata_file(self.g_pool.rec_dir, file_name)
+        annotations = pm.Mutable_Bisector(
+            annotation_pldata.data, annotation_pldata.timestamps
+        )
+        logger.info("Loaded {} annotations from recording.".format(len(annotations)))
+        return annotations
 
-    def export_annotations(self, export_range, export_dir):
+    def cleanup(self):
+        with fm.PLData_Writer(self.g_pool.rec_dir, "annotation_player") as writer:
+            for ts, annotation in zip(self.annotations.timestamps, self.annotations):
+                writer.append_serialized(ts, "annotation", annotation.serialized)
+
+    def customize_menu(self):
+        self.menu.label = "View and Edit Annotations"
+        self.menu.append(
+            ui.Info_Text(
+                "Annotations recorded with capture are displayed when this "
+                "plugin is loaded. New annotations can be added with the "
+                "interface below."
+            )
+        )
+        self.menu.append(
+            ui.Info_Text(
+                "If you want to revert annotations to the recorded state, "
+                "stop player, delete the annotation_player.pldata file in the recording "
+                "and reopen player."
+            )
+        )
 
-        if not self.annotations:
-            logger.warning('No annotations in this recording nothing to export')
+    def fire_annotation(self, annotation_label):
+        if self.last_frame_ts is None:
             return
-
-        export_window = pm.exact_window(self.g_pool.timestamps, export_range)
-        annotation_section = self.annotations.init_dict_for_window(export_window)
-        annotation_idc = pm.find_closest(self.g_pool.timestamps, annotation_section['data_ts'])
-        csv_keys = self.parse_csv_keys(annotation_section['data'])
-
-        with open(os.path.join(export_dir, 'annotations.csv'), 'w', encoding='utf-8', newline='') as csvfile:
-            csv_writer = csv.writer(csvfile)
-            csv_writer.writerow(csv_keys)
-            for annotation, idx in zip(annotation_section['data'], annotation_idc):
-                csv_row = [idx]
-                csv_row.extend((annotation.get(k, '') for k in csv_keys[1:]))
-                csv_writer.writerow(csv_row)
-            logger.info("Created 'annotations.csv' file.")
+        ts = self.last_frame_ts
+        logger.info('"{}"@{}'.format(annotation_label, ts))
```

Comment:
Why is this different than the log message in the Capture version in line 219?

Reply:
This is the annotation being fired, line 219 is the annotation being received. I think we should change this one to a debug message.

Reply:
Did you forget to push your changes here @ckbaumann ?

Reply:
No, just forgot to change this as well ;)